File: src/parser.cpp
#include "cardcode/parser.hpp"

#include "cardcode/lexer.hpp"

namespace cardcode {
namespace {

struct ParserState {
    const std::vector<Token>& tokens;
    std::size_t pos;
    std::vector<Diagnostic>& diags;

    const Token& peek() const { return tokens[pos]; }
    const Token& next() {
        const Token& t = tokens[pos];
        if (tokens[pos].kind != TokenKind::End) ++pos;
        return t;
    }
};

SourceSpan join(const SourceSpan& a, const SourceSpan& b) {
    return SourceSpan{a.start_offset, b.end_offset,
                      a.start_line, a.start_column,
                      b.end_line, b.end_column};
}

bool parse_expr(ParserState& s, RawExpr& out);

bool parse_list(ParserState& s, const Token& lparen, RawExpr& out) {
    out.kind = RawKind::List;
    SourceSpan span = lparen.span;

    while (true) {
        const Token& t = s.peek();
        if (t.kind == TokenKind::RParen) {
            span = join(span, t.span);
            s.next();
            out.span = span;
            return true;
        }
        if (t.kind == TokenKind::End) {
            s.diags.push_back({DiagnosticSeverity::Error, lparen.span,
                "missing closing parenthesis"});
            out.span = join(span, t.span);
            return false;
        }

        RawExpr child;
        if (!parse_expr(s, child)) {
            out.span = span;
            return false;
        }
        span = join(span, child.span);
        out.list.push_back(std::move(child));
    }
}

bool parse_expr(ParserState& s, RawExpr& out) {
    const Token& t = s.next();
    switch (t.kind) {
        case TokenKind::LParen:
            return parse_list(s, t, out);
        case TokenKind::Symbol:
            out.kind = RawKind::Symbol;
            out.symbol = t.text;
            out.span = t.span;
            return true;
        case TokenKind::Integer:
            out.kind = RawKind::Integer;
            out.integer = t.integer;
            out.span = t.span;
            return true;
        case TokenKind::RParen:
            s.diags.push_back({DiagnosticSeverity::Error, t.span, "unexpected ')'"});
            return false;
        case TokenKind::End:
            s.diags.push_back({DiagnosticSeverity::Error, t.span, "unexpected end of input"});
            return false;
    }
    return false;
}

} // namespace

ParseResult parse(std::string_view source) {
    ParseResult result;

    LexResult lexed = lex(source);
    result.diagnostics = lexed.diagnostics;

    ParserState s{lexed.tokens, 0, result.diagnostics};
    while (s.peek().kind != TokenKind::End) {
        RawExpr e;
        if (parse_expr(s, e)) {
            result.exprs.push_back(std::move(e));
        }
        // parse_expr always consumes at least one token, so the loop makes
        // progress and terminates even on error.
    }
    return result;
}

} // namespace cardcode

File: app/src/parser.js
export function parseCardCode(source) {
  const parser = new Parser(source);
  const children = [];
  while (!parser.atEnd()) {
    parser.skipIgnored();
    if (parser.atEnd()) break;
    children.push(parser.parseExpr());
  }
  return {
    type: 'program',
    children,
    span: spanFromOffsets(source, 0, source.length)
  };
}

class Parser {
  constructor(source) {
    this.source = source;
    this.offset = 0;
  }

  atEnd() {
    return this.offset >= this.source.length;
  }

  skipIgnored() {
    for (;;) {
      while (!this.atEnd() && /\s/.test(this.source[this.offset])) this.offset++;
      if (this.source[this.offset] !== ';') return;
      while (!this.atEnd() && this.source[this.offset] !== '\n') this.offset++;
    }
  }

  parseExpr() {
    this.skipIgnored();
    if (this.source[this.offset] === '(') return this.parseList();
    if (this.source[this.offset] === ')') throw this.error("unexpected ')'");
    return this.parseAtom();
  }

  parseList() {
    const start = this.offset++;
    const children = [];
    for (;;) {
      this.skipIgnored();
      if (this.atEnd()) throw this.error("expected ')'");
      if (this.source[this.offset] === ')') {
        this.offset++;
        return { type: 'list', children, span: spanFromOffsets(this.source, start, this.offset) };
      }
      children.push(this.parseExpr());
    }
  }

  parseAtom() {
    const start = this.offset;
    while (!this.atEnd() && !/\s|\(|\)|;/.test(this.source[this.offset])) this.offset++;
    const raw = this.source.slice(start, this.offset);
    const number = /^-?\d+$/.test(raw) ? Number(raw) : null;
    const value = number === null ? raw : number;
    const type = number === null ? 'symbol' : 'integer';
    return { type, value, raw, span: spanFromOffsets(this.source, start, this.offset) };
  }

  error(message) {
    const span = spanFromOffsets(this.source, this.offset, Math.min(this.source.length, this.offset + 1));
    const err = new Error(`${message} at ${span.startLine}:${span.startColumn}`);
    err.span = span;
    return err;
  }
}

export function spanFromOffsets(source, startOffset, endOffset) {
  const start = lineColumnAt(source, startOffset);
  const end = lineColumnAt(source, endOffset);
  return {
    startOffset,
    endOffset,
    startLine: start.line,
    startColumn: start.column,
    endLine: end.line,
    endColumn: end.column
  };
}

function lineColumnAt(source, offset) {
  let line = 1;
  let column = 1;
  for (let i = 0; i < offset; i++) {
    if (source[i] === '\n') {
      line++;
      column = 1;
    } else {
      column++;
    }
  }
  return { line, column };
}
